File: backend/src/main/java/com/fincoach/controller/ProfileController.java
package com.fincoach.controller;

import com.fincoach.dto.FinancialProfileRequest;
import com.fincoach.dto.ScoreResult;
import com.fincoach.model.FinancialProfile;
import com.fincoach.repository.FinancialProfileRepository;
import com.fincoach.service.ActionPlanService;
import com.fincoach.service.FinancialScoringService;
import jakarta.validation.Valid;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.fincoach.util.NumberUtils.orZero;

@RestController
@RequestMapping("/api")
public class ProfileController {

    private static final Logger log = LoggerFactory.getLogger(ProfileController.class);

    private final FinancialProfileRepository profileRepo;
    private final FinancialScoringService scoringService;
    private final ActionPlanService actionPlanService;

    public ProfileController(FinancialProfileRepository profileRepo,
                             FinancialScoringService scoringService,
                             ActionPlanService actionPlanService) {
        this.profileRepo = profileRepo;
        this.scoringService = scoringService;
        this.actionPlanService = actionPlanService;
    }

    /** GET /api/profile — retourne le profil de l'utilisateur connecté. */
    @GetMapping("/profile")
    public ResponseEntity<?> getMyProfile(@AuthenticationPrincipal String userId) {
        log.info("GET /profile for userId={}", userId);
        Optional<FinancialProfile> profileOpt = profileRepo.findTopByUserIdOrderByUpdatedAtDesc(userId);
        if (profileOpt.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(toResponse(profileOpt.get()));
    }

    /** GET /api/profile/{userId} — conservé pour compatibilité dashboard. */
    @GetMapping("/profile/{userId}")
    public ResponseEntity<?> getProfile(@PathVariable String userId) {
        log.info("GET /profile/{}", userId);
        Optional<FinancialProfile> profileOpt = profileRepo.findTopByUserIdOrderByUpdatedAtDesc(userId);
        if (profileOpt.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(toResponse(profileOpt.get()));
    }

    /**
     * POST /api/profile — upsert du profil financier.
     * L'userId est extrait du JWT (jamais du body), ce qui empêche toute usurpation.
     */
    @PostMapping("/profile")
    public ResponseEntity<?> saveProfile(
            @AuthenticationPrincipal String userId,
            @Valid @RequestBody FinancialProfileRequest req,
            BindingResult bindingResult) {

        if (bindingResult.hasErrors()) {
            Map<String, String> errors = bindingResult.getFieldErrors().stream()
                    .collect(Collectors.toMap(
                            fe -> fe.getField(),
                            fe -> fe.getDefaultMessage(),
                            (a, b) -> a));
            log.warn("Validation errors for userId={}: {}", userId, errors);
            return ResponseEntity.badRequest().body(Map.of("errors", errors));
        }

        log.info("POST /profile (upsert) for userId={}", userId);

        // Upsert : mise à jour si profil existant, création sinon
        Optional<FinancialProfile> existingOpt = profileRepo.findTopByUserIdOrderByUpdatedAtDesc(userId);
        FinancialProfile profile = existingOpt.orElse(new FinancialProfile());
        profile.setUserId(userId);

        profile.setMonthlyIncome(req.getMonthlyIncome());
        profile.setOtherIncome(req.getOtherIncome() != null ? req.getOtherIncome() : 0.0);
        profile.setRent(req.getRent());
        profile.setUtilities(req.getUtilities() != null ? req.getUtilities() : 0.0);
        profile.setInsurance(req.getInsurance() != null ? req.getInsurance() : 0.0);
        profile.setLoans(req.getLoans());
        profile.setSubscriptions(req.getSubscriptions() != null ? req.getSubscriptions() : 0.0);
        profile.setFood(req.getFood());
        profile.setTransport(req.getTransport() != null ? req.getTransport() : 0.0);
        profile.setLeisure(req.getLeisure() != null ? req.getLeisure() : 0.0);
        profile.setClothing(req.getClothing() != null ? req.getClothing() : 0.0);
        profile.setHealth(req.getHealth() != null ? req.getHealth() : 0.0);
        profile.setCurrentSavings(req.getCurrentSavings());
        profile.setTotalDebt(req.getTotalDebt());
        profile.setMonthlySavingsGoal(req.getMonthlySavingsGoal() != null ? req.getMonthlySavingsGoal() : 0.0);
        profile.setTypeHabitation(req.getTypeHabitation());
        profile.setSituationFamiliale(req.getSituationFamiliale());
        profile.setNombrePersonnes(req.getNombrePersonnes() != null ? req.getNombrePersonnes() : 0);

        scoringService.computeScores(profile);
        FinancialProfile saved = profileRepo.save(profile);
        log.info("Profile upserted id={}, score={}", saved.getId(), saved.getFinancialScore());
        actionPlanService.generateFromProfile(saved);

        return ResponseEntity.ok(toResponse(saved));
    }

    @GetMapping("/dashboard/{userId}")
    public ResponseEntity<Map<String, Object>> getDashboard(@PathVariable String userId) {
        log.info("GET /dashboard/{}", userId);
        Map<String, Object> dashboard = new LinkedHashMap<>();

        Optional<FinancialProfile> profileOpt = profileRepo.findTopByUserIdOrderByUpdatedAtDesc(userId);
        dashboard.put("profile", profileOpt.map(this::toResponse).orElse(null));

        var activeActions = actionPlanService.getActionsForUser(userId).stream()
                .filter(a -> com.fincoach.model.ActionStatut.EN_COURS == a.getStatus())
                .limit(4)
                .map(actionPlanService::toResponse)
                .toList();
        dashboard.put("topActions", activeActions);
        dashboard.put("stats", actionPlanService.buildStats(userId));

        return ResponseEntity.ok(dashboard);
    }

    Map<String, Object> toResponse(FinancialProfile p) {
        double income = orZero(p.getMonthlyIncome()) + orZero(p.getOtherIncome());
        double fixed = orZero(p.getRent()) + orZero(p.getUtilities()) + orZero(p.getInsurance())
                + orZero(p.getLoans()) + orZero(p.getSubscriptions());
        double variable = orZero(p.getFood()) + orZero(p.getTransport()) + orZero(p.getLeisure())
                + orZero(p.getClothing()) + orZero(p.getHealth());
        double total = fixed + variable;

        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", p.getId());
        m.put("userId", p.getUserId());
        m.put("monthlyIncome", income);
        m.put("totalFixedExpenses", fixed);
        m.put("totalVariableExpenses", variable);
        m.put("totalExpenses", total);
        m.put("monthlySurplus", income - total);
        m.put("savingsRate", p.getSavingsRate());
        m.put("debtRatio", p.getDebtRatio());
        m.put("financialScore", p.getFinancialScore());
        m.put("currentSavings", p.getCurrentSavings());
        m.put("totalDebt", p.getTotalDebt());
        m.put("typeHabitation", p.getTypeHabitation());
        m.put("situationFamiliale", p.getSituationFamiliale());
        m.put("nombrePersonnes", p.getNombrePersonnes());
        m.put("insights", scoringService.generateInsights(p));
        m.put("breakdown", Map.of(
                "housing", orZero(p.getRent()),
                "food", orZero(p.getFood()),
                "transport", orZero(p.getTransport()),
                "leisure", orZero(p.getLeisure()),
                "loans", orZero(p.getLoans()),
                "other", orZero(p.getUtilities()) + orZero(p.getInsurance())
                        + orZero(p.getSubscriptions()) + orZero(p.getClothing()) + orZero(p.getHealth())));
        m.put("updatedAt", p.getUpdatedAt());

        // Score breakdown (calculé à la volée — non persisté, omis si profil incomplet)
        ScoreResult scoreResult = scoringService.calculateScore(p);
        if (!"N/A".equals(scoreResult.getGrade())) {
            m.put("scoreBreakdown", Map.of(
                    "grade", scoreResult.getGrade(),
                    "totalScore", scoreResult.getTotalScore(),
                    "breakdown", scoreResult.getBreakdown(),
                    "message", scoreResult.getMessage()
            ));
        }

        // Champs bruts pour pré-remplissage du formulaire
        m.put("raw", Map.ofEntries(
                Map.entry("monthlyIncome", orZero(p.getMonthlyIncome())),
                Map.entry("otherIncome", orZero(p.getOtherIncome())),
                Map.entry("rent", orZero(p.getRent())),
                Map.entry("utilities", orZero(p.getUtilities())),
                Map.entry("insurance", orZero(p.getInsurance())),
                Map.entry("loans", orZero(p.getLoans())),
                Map.entry("subscriptions", orZero(p.getSubscriptions())),
                Map.entry("food", orZero(p.getFood())),
                Map.entry("transport", orZero(p.getTransport())),
                Map.entry("leisure", orZero(p.getLeisure())),
                Map.entry("clothing", orZero(p.getClothing())),
                Map.entry("health", orZero(p.getHealth())),
                Map.entry("currentSavings", orZero(p.getCurrentSavings())),
                Map.entry("totalDebt", orZero(p.getTotalDebt())),
                Map.entry("monthlySavingsGoal", orZero(p.getMonthlySavingsGoal()))
        ));
        return m;
    }
}

File: backend/src/controllers/profile.controller.ts
import { Response } from 'express';
import { AuthRequest, FinancialProfilePayload } from '../types';
import prisma from '../config/database';
import { scoringService } from '../services/scoring.service';
import { actionPlanService } from '../services/action-plan.service';
import { FinancialProfile } from '@prisma/client';
import { createLogger } from '../utils/logger';

const logger = createLogger('ProfileController');

/**
 * Profile controller — port of Java ProfileController.java.
 *
 * Routes:
 *  GET  /api/profile              — get current user's profile (from JWT)
 *  GET  /api/profile/:userId      — get profile by userId (backward compat)
 *  POST /api/profile              — upsert financial profile
 *  GET  /api/dashboard/:userId    — dashboard (profile + top actions + stats)
 */
export const profileController = {
  // GET /api/profile
  async getMyProfile(req: AuthRequest, res: Response): Promise<void> {
    logger.debug('GET /api/profile', { userId: req.userId });

    const profile = await prisma.financialProfile.findFirst({
      where: { userId: req.userId },
      orderBy: { updatedAt: 'desc' },
    });

    if (!profile) {
      logger.warn('Profile not found', { userId: req.userId });
      res.status(404).json({ error: 'Profil introuvable.' });
      return;
    }

    logger.debug('Profile fetched', { userId: req.userId, profileId: profile.id });
    res.json(buildProfileResponse(profile));
  },

  // GET /api/profile/:userId
  async getProfileByUserId(req: AuthRequest, res: Response): Promise<void> {
    const { userId } = req.params as { userId: string };
    logger.debug('GET /api/profile/:userId', { requestedUserId: userId, callerUserId: req.userId });

    const profile = await prisma.financialProfile.findFirst({
      where: { userId },
      orderBy: { updatedAt: 'desc' },
    });

    if (!profile) {
      logger.warn('Profile not found by userId', { requestedUserId: userId });
      res.status(404).json({ error: 'Profil introuvable.' });
      return;
    }

    logger.debug('Profile fetched by userId', { requestedUserId: userId, profileId: profile.id });
    res.json(buildProfileResponse(profile));
  },

  // POST /api/profile
  async saveProfile(req: AuthRequest, res: Response): Promise<void> {
    const userId = req.userId;
    const body = req.body as FinancialProfilePayload;

    logger.info('POST /api/profile — saving financial profile', {
      userId,
      monthlyIncome: body.monthlyIncome,
    });

    if (body.monthlyIncome === undefined || body.monthlyIncome === null) {
      logger.warn('Save profile validation failed — missing monthlyIncome', { userId });
      res.status(400).json({ error: 'Le revenu mensuel est requis.' });
      return;
    }

    // Upsert: update existing profile or create a new one
    const existing = await prisma.financialProfile.findFirst({
      where: { userId },
      orderBy: { updatedAt: 'desc' },
    });

    const orZero = (v: number | undefined | null) => v ?? 0;

    // Compute scores from the incoming data
    const profileData = {
      monthlyIncome: orZero(body.monthlyIncome),
      otherIncome: orZero(body.otherIncome),
      rent: orZero(body.rent),
      utilities: orZero(body.utilities),
      insurance: orZero(body.insurance),
      loans: orZero(body.loans),
      subscriptions: orZero(body.subscriptions),
      food: orZero(body.food),
      transport: orZero(body.transport),
      leisure: orZero(body.leisure),
      clothing: orZero(body.clothing),
      health: orZero(body.health),
      currentSavings: orZero(body.currentSavings),
      totalDebt: orZero(body.totalDebt),
      monthlySavingsGoal: orZero(body.monthlySavingsGoal),
      typeHabitation: body.typeHabitation ?? null,
      situationFamiliale: body.situationFamiliale ?? null,
      nombrePersonnes: body.nombrePersonnes ?? 0,
    };

    const scores = scoringService.computeScores(profileData);

    const savedProfile = existing
      ? await prisma.financialProfile.update({
          where: { id: existing.id },
          data: { ...profileData, ...scores },
        })
      : await prisma.financialProfile.create({
          data: { userId, ...profileData, ...scores },
        });

    logger.info('Profile upserted successfully', {
      userId,
      profileId: savedProfile.id,
      financialScore: savedProfile.financialScore,
      isUpdate: !!existing,
    });

    // Auto-generate action plans based on the new profile
    await actionPlanService.generateFromProfile(savedProfile);

    res.json(buildProfileResponse(savedProfile));
  },

  // GET /api/dashboard/:userId
  async getDashboard(req: AuthRequest, res: Response): Promise<void> {
    const { userId } = req.params as { userId: string };
    logger.debug('GET /api/dashboard/:userId', { requestedUserId: userId, callerUserId: req.userId });

    const profile = await prisma.financialProfile.findFirst({
      where: { userId },
      orderBy: { updatedAt: 'desc' },
    });

    const allActions = await actionPlanService.getActionsForUser(userId);
    const topActions = allActions
      .filter((a) => a.status === 'EN_COURS')
      .slice(0, 4)
      .map(actionPlanService.toResponse.bind(actionPlanService));

    const stats = await actionPlanService.buildStats(userId);

    logger.debug('Dashboard data loaded', {
      userId,
      hasProfile: !!profile,
      topActionsCount: topActions.length,
    });

    res.json({
      profile: profile ? buildProfileResponse(profile) : null,
      topActions,
      stats,
    });
  },
};

// ─── Profile response builder ──────────────────────────────────────────────────

function buildProfileResponse(p: FinancialProfile): Record<string, unknown> {
  const orZero = (v: number | null | undefined) => v ?? 0;

  const income = orZero(p.monthlyIncome) + orZero(p.otherIncome);
  const fixed =
    orZero(p.rent) +
    orZero(p.utilities) +
    orZero(p.insurance) +
    orZero(p.loans) +
    orZero(p.subscriptions);
  const variable =
    orZero(p.food) +
    orZero(p.transport) +
    orZero(p.leisure) +
    orZero(p.clothing) +
    orZero(p.health);
  const total = fixed + variable;

  const scoreResult = scoringService.calculateScore(p);

  const response: Record<string, unknown> = {
    id: p.id.toString(),
    userId: p.userId,
    monthlyIncome: income,
    totalFixedExpenses: fixed,
    totalVariableExpenses: variable,
    totalExpenses: total,
    monthlySurplus: income - total,
    savingsRate: p.savingsRate,
    debtRatio: p.debtRatio,
    financialScore: p.financialScore,
    currentSavings: p.currentSavings,
    totalDebt: p.totalDebt,
    typeHabitation: p.typeHabitation,
    situationFamiliale: p.situationFamiliale,
    nombrePersonnes: p.nombrePersonnes,
    insights: scoringService.generateInsights(p),
    breakdown: {
      housing: orZero(p.rent),
      food: orZero(p.food),
      transport: orZero(p.transport),
      leisure: orZero(p.leisure),
      loans: orZero(p.loans),
      other:
        orZero(p.utilities) +
        orZero(p.insurance) +
        orZero(p.subscriptions) +
        orZero(p.clothing) +
        orZero(p.health),
    },
    updatedAt: p.updatedAt,
    raw: {
      monthlyIncome: orZero(p.monthlyIncome),
      otherIncome: orZero(p.otherIncome),
      rent: orZero(p.rent),
      utilities: orZero(p.utilities),
      insurance: orZero(p.insurance),
      loans: orZero(p.loans),
      subscriptions: orZero(p.subscriptions),
      food: orZero(p.food),
      transport: orZero(p.transport),
      leisure: orZero(p.leisure),
      clothing: orZero(p.clothing),
      health: orZero(p.health),
      currentSavings: orZero(p.currentSavings),
      totalDebt: orZero(p.totalDebt),
      monthlySavingsGoal: orZero(p.monthlySavingsGoal),
    },
  };

  if (scoreResult.grade !== 'N/A') {
    response['scoreBreakdown'] = {
      grade: scoreResult.grade,
      totalScore: scoreResult.totalScore,
      breakdown: scoreResult.breakdown,
      message: scoreResult.message,
    };
  }

  return response;
}
